L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,236 @@
+
+2: It realizes TLS-related features such as credential reload and server authorization check in a synchronous manner. It is undesirable in SPIFFE use case as those operations could involve expensive RPC calls to other services which may block gRPC core threads.
+
+3: The current TLS credential API is inflexible in incorporating new TLS-related features in future as not only does it break the API, it also requires all wrapped languages to update their implementations in a lockstep manner.
+
+This proposal introduces a new credential API model that allows flexible addition of TLS-related features that can be implemented in either a synchronous or asynchronous manner. It then proposes new credential API's for both HTTPS and SPIFFE-based TLS that complies with the API model.
+
+### Related Proposals:
+N/A
+
+## Proposal
+
+The first part of proposal introduces credential options and credential create APIs that will be added to include/grpc/grpc_security.h. Note that although the credentials options provided to HTTPS and SPIFFE-based TLS credential API have the same format (i.e., grpc_tls_credentials_options), some of its members will not get populated and hence used in some API's, for instance, grpc_tls_server_authorization_check_config will be ignored in server-side API's.
+
+```c++
+
+/** Config for TLS key materials. */
+typedef struct grpc_tls_key_materials_config grpc_tls_key_materials_config;
+
+/** Config for TLS credential reload. */
+typedef struct grpc_tls_credential_reload_config
+    grpc_tls_credential_reload_config;
+
+/** Config for TLS server authorization check. */
+typedef struct grpc_tls_server_authorization_check_config
+    grpc_tls_server_authorization_check_config;
+
+/** TLS credentials options. */
+typedef struct grpc_tls_credentials_options grpc_tls_credentials_options;
+
+/** Create an empty TLS credentials options. */
+GRPCAPI grpc_tls_credentials_options* grpc_tls_credentials_options_create();
+
+/** Set grpc_ssl_client_certificate_request_type field in credentials options
+    with the provided type. options should not be NULL.
+    It returns 1 on success and 0 on failure. */
+GRPCAPI int grpc_tls_credentials_options_set_cert_request_type(
+    grpc_tls_credentials_options* options,
+    grpc_ssl_client_certificate_request_type type);
+
+/** Set grpc_tls_key_materials_config field in credentials options
+    with the provided config struct whose ownership is transferred.
+    Both parameters should not be NULL.
```

Comment:
clarity: `Both parameters should not be NULL.` is ambiquous.  one of them needs to be set? setting just one of them to null is illegal?  Also "should" is not the right word here I think?